src/drivers/intel/ssp/ssp.c

```diff
@@ -246,8 +246,17 @@ static int ssp_set_config_tplg(struct dai *dai, struct ipc_config_dai *common_co
 	/* ignore config if SSP is already configured */
 	if (ssp->state[DAI_DIR_PLAYBACK] > COMP_STATE_READY ||
 	    ssp->state[DAI_DIR_CAPTURE] > COMP_STATE_READY) {
-		dai_info(dai, "ssp_set_config(): Already configured. Ignore config");
-		goto clk;
+		if (!memcmp(&ssp->params, &config->ssp, sizeof(ssp->params))) {
+			dai_info(dai, "ssp_set_config(): Already configured. Ignore config");
+			goto clk;
+		}
+
+		if (ssp->state[DAI_DIR_PLAYBACK] == COMP_STATE_ACTIVE ||
+		    ssp->state[DAI_DIR_CAPTURE] == COMP_STATE_ACTIVE) {
+			dai_err(dai, "ssp_set_config(): New config but DAI running");
+			ret = -EBUSY;
+			goto out;
+		}
```

Comment:
I don't think this works in the case of the MCLK_ES and BCLK_ES cases @kv2019i 

the SSP will be BCLK_ACTIVE and you can't reprogram it.

you would need to have an additional check on BCLK_ACTIVE and MCLK_ACTIVE.

Reply:
@plbossart But this would have happened before this patch as well, right? If state is greater than READY, old code jumped to "goto clk" always, so the clocks were always reprogrammed. Hopefully @brentlu can chime in, you made a patch to this code some time back.

The check on L254 should always be an error. If the DAI config is indeed changed, but the stream is running, this is an error. This was not ignored before, but now that I check whether config was changed, I can catch this in the driver.

Reply:
@kv2019i the routines for ssp_mclk_prepare_enable () and ssp_bclk_prepare_enable() check the MCLK_ and BCLK_ACTIVE states, so the clocks wouldn't be re-enabled twice.

Reply:
@plbossart  Ok. Not sure I still get all the dependencies here, but let me change this so that we go to clock code always. New version to be uploaded shortly.

Reply:
The problematic case is if the state is COMP_STATE_PREPARED but the BCLK_ and MCLK are active. That's where you cannot reconfigure the SSP because in practice it's already toggling.


The COMP_STATE_ACTIVE means data transfers are happening with a trigger, that is different to the early clock enablement where there is no data transfers yet.

Reply:
@plbossart Oh man, how can a small 570 line function be so complicated. ;) 

But yes, this would cause problems. Uploaded yet one more versions that checks for the case when BCLK and MCLK are active in PREPARED state. And if yes, hw reconfig is not done.

Basicly SSP multi-config (as used for Bluetooth offload) is incompatible with early MCLK/BCLK, with our current interfaces. This can be worked around by moving to dynamic pipelines (which just avoids the problem as SSP is not configured until the Bluetooth configuration is known at PCM open/hw_params time).

Reply:
@kv2019i I agree multiple configurations should not be used with BCLK_ES. This was not designed for it.

But maybe lateral thinking: why can't we fix the topology to remove BCLK_ES for the SSP used for offload?

Reply:
@plbossart That's not the problem with BT. The multi-config mechanism is simply broken. When card is probed, we set up SSP with the static config in topology. When PCM is opened, we need to change the configure and update SSP to different parameters. This now fails due to current code in ssp.c that prevent any changes if state is above READY. 

This PR fixes BT offload, but it leaves a limitation that BCLK_ES/MCLK_ES cannot be used with BT-offload. This is not the case now, so not a problem currently.

So we really do need this fix.

Reply:
Multi-config and early clock could happen at the same time when sharing one I2S port with multiple codecs but so far we don't have this kind of design on Chromebooks. If this happens, we need to restart mclk/bclk with new frequency instead of a warning message. The code below just reconfigure the SSP port data format.

I think the patch won't break the early clock mechanism so long as we don't enable multi-config and early clock together.